inference.py

```diff
@@ -16,133 +16,117 @@
-                            if index == 0:
-                                tqdm.write(f'(debug mode) Visualizing inferred tiles...')
-                            vis_from_batch(params, inputs, outputs, batch_index=0, vis_path=output_path,
-                                           dataset=f'{row_start}_{col_start}_inf', ep_num=index, debug=True)
-
-                        outputs = F.softmax(outputs, dim=1)
-                        output_counts[row_start:row_end, col_start:col_end] += 1
-
-                        # Add inference on sub-image to all completed inferences on previous sub-images.
-                        # TODO: This operation need to be optimized. Using a lot of RAM on large images.
-                        output_probs[:, row_start:row_end, col_start:col_end] += np.squeeze(outputs.cpu().numpy(),
-                                                                                            axis=0)
-
-                        if debug and device.type == 'cuda':
-                            res, mem = gpu_stats(device=device.index)
-                            _tqdm.set_postfix(OrderedDict(gpu_perc=f'{res.gpu} %',
-                                                          gpu_RAM=f'{mem.used / (1024 ** 2):.0f}/{mem.total / (1024 ** 2):.0f} MiB',
-                                                          inp_size=inputs.cpu().numpy().shape,
-                                                          out_size=outputs.cpu().numpy().shape,
-                                                          overlay=overlay))
-
-            # Divide array according to output counts. Manages overlap and returns a softmax array as if only one forward pass had been done.
-            output_mask_raw = np.divide(output_probs, np.maximum(output_counts, 1))  # , 1 is added to overwrite 0 values.
-
-            # Resize the output array to the size of the input image and write it
-            output_mask_raw_cropped = np.moveaxis(output_mask_raw, 0, -1)
-            output_mask_raw_cropped = output_mask_raw_cropped[overlay:(h + overlay), overlay:(w + overlay), :]
-
-            return output_mask_raw_cropped
-    else:
-        raise IOError(f"Error classifying image : Image shape of {len(nd_array.shape)} is not recognized")
-
+    img_array, input_image, dataset_nodata = image_reader_as_array(input_image=raster,
+                                                                   clip_gpkg=clip_gpkg)
+    metadata = add_metadata_from_raster_to_sample(img_array, 
+                                                    input_image, 
+                                                    meta_map=None, 
+                                                    raster_info=None)
+    h, w, bands = img_array.shape
+    assert num_bands <= bands, f"Num of specified bands is not compatible with image shape {img_array.shape}"
+    if num_bands < bands:
+       img_array = img_array[:, :, :num_bands]
+    
+    padding = int(round(sample_size * (1 - 1.0/overlap)))
```

Comment:
hmm, I don't understand this. If your overlap is set to "2" as default, you're actually padding with 50% of the sample size. Are you sure you didn't mean:
dist_samples = int(round(sample_size * (1 - (overlap / 1.0)))
If so, then your considering the overlap as a ratio, not a number of pixels. Still, the default value of "2" wouldn't make sense.
I would suggest keeping it like this, but having it as a percentage:
dist_samples = round(sample_size * (1 - (overlap / 100)))
(the "int" conversion isn't necessary, since "round" already outputs an int...)
the default value could be 25 (ie 25%), for example.

Reply:
How the padding is calculated in this case is slightly different, the intention would be to use a 'valid' border padding strategy instead of 'same' padding with zeros. In this case, np.pad(mode='reflect') can be used.  https://numpy.org/doc/stable/reference/generated/numpy.pad.html. Because of this, there are hard limits set on the overlap: In the README.md:  'overlap between tiles for smoothing. Must be an even number that divides chunk_size without remainder'. Also round does not always output int, take a look at this official doc on round in Python3 https://docs.python.org/3/library/functions.html#round

I understand the confusion with the overlap param and how it clashes with the one set in 'sample'. Perhaps I could change the name of this param to avoid such.